### PR TITLE
Feat/multipass

### PR DESCRIPTION
**Pull request by Guillaumebeuzeboc**

### CODE REVIEW

File: colcon_in_container/providers/multipass.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
+# Copyright (C) 2023 Canonical, Ltd.
```
Comment (artivis) [inline colcon_in_container/providers/multipass.py:1]:
```suggestion
# Copyright (C) 2024 Canonical, Ltd.
```

File: colcon_in_container/providers/multipass.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        return subprocess.run(command, **kwargs)
+
+    def _clean_instance(self):
+        self._run(['delete', '--purge', self.instance_name])
+
+    def execute_command(self, command: List[str]):
+        """Execute the given command inside the instance."""
+        completed_process = self._run(['exec', self.instance_name,
+                                       '--working-directory', '/ws',
+                                       '--', 'sudo', *command],
```
Comment (artivis) [inline colcon_in_container/providers/multipass.py:104]: 'sudo' ?

Comment (Guillaumebeuzeboc) [reply colcon_in_container/providers/multipass.py:104]: lxd containers spawn only with root user.
Multipass spawn with root and ubuntu user. By default, everything executed as the `ubuntu` user.
I decided to "ignore" the ubuntu user on multipass by running everything as root in the vm.

Comment (artivis) [reply colcon_in_container/providers/multipass.py:104]: Alright, we should try to use the regular user as much as possible. Let's leave it a that here but please do open a ticket for later.

File: colcon_in_container/providers/multipass.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                'Multipass is not installed.'
+                'Please run `sudo snap install multipass`')
+
+        self._render_jinja_template()
+
+        if self._run(['info', self.instance_name]).returncode == 0:
+            self._clean_instance()
+
+        cpus = os.getenv('COLCON_IN_CONTAINER_MULTIPASS_CPUS', default='2')
+        mem = os.getenv('COLCON_IN_CONTAINER_MULTIPASS_MEMORY', default='2G')
```
Comment (artivis) [inline colcon_in_container/providers/multipass.py:50]: The default ain't much for compiling stuff, maybe default to 4G?

File: colcon_in_container/providers/multipass.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        disk = os.getenv('COLCON_IN_CONTAINER_MULTIPASS_DISK', default='256G')
```
Comment (artivis) [inline colcon_in_container/providers/multipass.py:51]: This on the other hand is way to much :D ! How about defaulting to e.g. 16G ?

Comment (Guillaumebeuzeboc) [reply colcon_in_container/providers/multipass.py:51]: I took the default values from craft provider: https://github.com/canonical/craft-providers/blob/b0ada6a7410575550a16ee03521eabf1ec2d7751/craft_providers/multipass/multipass_instance.py#L337

Multipass won't take all that space on the disk if not necessary.

File: colcon_in_container/providers/multipass.py (hunk 3 above)
Comment (Guillaumebeuzeboc) [reply colcon_in_container/providers/multipass.py:50]: OK I will change that.

File: colcon_in_container/providers/multipass.py (hunk 2 above)
Comment (Guillaumebeuzeboc) [reply colcon_in_container/providers/multipass.py:104]: Ticket created: https://github.com/canonical/colcon-in-container/issues/18

**Review by artivis — APPROVED** [review]

**Pull request merged**